### PR TITLE
feat(FEC-9227): support reInit of MSE

### PR DESCRIPTION
**Pull request by Yuvalke**
### Description of the Changes

Destroy and initialize hls.js before and after ads for now(should be resolved by hls.js later and it'll be replace by attach detach media)

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/hls-adapter.js [hunk 1]
```diff
@@ -608,12 +658,14 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
    * @private
    */
   _getLevelDetails(): Object {
-    const level =
-      this._hls.levels[this._hls.currentLevel] ||
-      this._hls.levels[this._hls.nextLevel] ||
-      this._hls.levels[this._hls.nextAutoLevel] ||
-      this._hls.levels[this._hls.nextLoadLevel];
-    return level && level.details ? level.details : {};
+    if (this._hls) {
```
Comment (OrenMe) [inline src/hls-adapter.js]: you must return an Object

File: src/hls-adapter.js [hunk 2]
```diff
@@ -598,7 +646,9 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
    * @public
    */
   isAdaptiveBitrateEnabled(): boolean {
-    return this._hls.autoLevelEnabled;
+    if (this._hls) {
```
Comment (OrenMe) [inline src/hls-adapter.js]: you must return a boolean

File: src/hls-adapter.js [hunk 3]
```diff
@@ -524,7 +572,7 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
    */
   selectTextTrack(textTrack: TextTrack): void {
     if (textTrack instanceof TextTrack && !textTrack.active) {
-      if (this._hls.subtitleTracks.length) {
+      if (this._hls && this._hls.subtitleTracks.length) {
```
Comment (OrenMe) [inline src/hls-adapter.js]: this is incorrect, if there's no this._hls you need to do nothing, not to go to the else clause

File: src/hls-adapter.js [hunk 4]
```diff
@@ -571,7 +619,7 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
    * @public
    */
   hideTextTrack(): void {
-    if (this._hls.subtitleTracks.length) {
+    if (this._hls && this._hls.subtitleTracks.length) {
```
Comment (OrenMe) [inline src/hls-adapter.js]: same comment as `selectTextTrack`

File: src/hls-adapter.js [hunk 5]
```diff
@@ -276,7 +289,42 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
+  /**
+   * detach media - will remove the media source from handling the video
+   * @public
+   * @returns {void}
+   */
+  detachMediaSource(): void {
+    if (this._hls) {
+      this._lastTimeDetach = this.currentTime;
+      this._reset();
+      this._hls = null;
```
[the remark this answers is not in the document]
Comment (Yuvalke) [reply src/hls-adapter.js]: I don't think that loadPromise should be clear on reset, it'll be fulfilled only on load and in the other cases we are not calling load

File: src/hls-adapter.js [hunk 6]
```diff
@@ -276,7 +289,42 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
+        Utils.Dom.removeAttribute(this._videoElement, 'src');
+      }
+      this._init(this._config);
+      if (!isNaN(this._lastTimeDetach) && !playbackEnded) {
+        const canPlayHandler = () => {
+          this._videoElement.removeEventListener(Html5EventType.CAN_PLAY, canPlayHandler);
+          this.currentTime = this._lastTimeDetach;
+          this._lastTimeDetach = NaN;
+        };
+        this._videoElement.addEventListener(Html5EventType.CAN_PLAY, canPlayHandler);
```
[the remark this answers is not in the document]
Comment (Yuvalke) [reply src/hls-adapter.js]: Do you think it better to create event manager just for this case?

File: src/hls-adapter.js [hunk 7]
```diff
@@ -276,7 +289,41 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
       }
     }
   }
-
+  /**
+   * attach media - return the media source to handle the video tag
+   * @public
+   * @param {boolean} playbackEnded don't seek to the last detach point
```
Comment (yairans) [inline src/hls-adapter.js]: fix

File: src/hls-adapter.js [hunk 8]
```diff
@@ -276,7 +289,41 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
+   * @returns {void}
+   */
+  attachMediaSource(playbackEnded: ?boolean): void {
+    if (!this._hls) {
+      if (this._videoElement && this._videoElement.src) {
+        Utils.Dom.setAttribute(this._videoElement, 'src', '');
+        Utils.Dom.removeAttribute(this._videoElement, 'src');
```
Comment (yairans) [inline src/hls-adapter.js]: can you explain these 2 lines?

Comment (Yuvalke) [reply src/hls-adapter.js]: Ima on some cases keep empty src, and it made errors

Comment (Yuvalke) [reply src/hls-adapter.js]: Ima on some cases keep empty src, and it made errors

Comment (yairans) [reply src/hls-adapter.js]: so just remove

Comment (Yuvalke) [reply src/hls-adapter.js]: I'm not sure why but if I don't do that it, it's not working..

Comment (Yuvalke) [reply src/hls-adapter.js]: It could happened when IMA doesn't clear the src on ADBREAKEND or didn't finish the clear

**Pull request merged**